pytensor/tensor/rewriting/linalg.py

```diff
@@ -377,3 +382,47 @@ def local_lift_through_linalg(
                 return [block_diag(*inner_matrices)]
             else:
                 raise NotImplementedError  # pragma: no cover
+
+
+@register_canonicalize
+@register_stabilize
+@register_specialize
+@node_rewriter([SVD])
+def local_svd_uv_simplify(fgraph, node):
+    """If we have more than one `SVD` `Op`s and at least one has keyword argument
+    `compute_uv=True`, then we can change `compute_uv = False` to `True` everywhere
+    and allow `pytensor` to re-use the decomposition outputs instead of recomputing.
+    """
+    (x,) = node.inputs
+
+    if node.compute_uv:
+        # compute_uv=True returns [u, s, v].
+        # if at least u or v is used, no need to rewrite this node.
+        if (
+            fgraph.clients[node.outputs[0]] is not None
+            or fgraph.clients[node.outputs[2]] is not None
+        ):
+            return
+
+        # Else, has to replace the s of this node with s of an SVD Op that compute_uv=False.
+        # First, iterate to see if there is an SVD Op that can be reused.
+        for cl, _ in fgraph.clients[x]:
+            if cl == "output":
+                continue
+            if isinstance(cl.op, Blockwise) and isinstance(cl.op.core_op, SVD):
+                if not cl.op.core_op.compute_uv:
+                    return {fgraph.clients[node.outputs[1]]: cl.outputs[0]}
+
+        # If no SVD reusable, return a new one.
+        return [svd(x, full_matrices=node.full_matrices, compute_uv=False)]
+
+    else:
+        # compute_uv=False returns [s].
+        # We want rewrite if there is another one with compute_uv=True.
+        # For this case, just reuse the `s` from the one with compute_uv=True.
+        for cl, _ in fgraph.clients[x]:
+            if cl == "output":
+                continue
+            if isinstance(cl.op, Blockwise) and isinstance(cl.op.core_op, SVD):
+                if cl.op.core_op.compute_uv:
+                    return [cl.outputs[1]]
```

Comment:
Thanks @ricardoV94. My understanding is like this: The `SVD` with `compute_uv == False` will return `[s]`, while the one with `compute_uv == True` will return `[u, s, v]`. We want to rewrite when there are 2 `SVD` `Op`s using the same input in the graph with different `compute_uv` value. Let's take the specific example of 2 `SVD` `Op`s, `svd_f` which returns `[s_f]` and `svd_t` which returns `[u_t, s_t, v_t]`. Based on whether _at least_ `u_t` or `v_t` is used (since we still have to calculate both even if we use just one of them for subsequent calculations), 1 of 2 rewrites can happen:

- Case 1: If at least `u_t` or `v_t` is used: return `[s_t]` in place of `[s_f]`.
- Case 2: Else: return `[s_f]` in place of `[s_t]`.
- Case 3: Additionally, if there is just one `SVD` `Op` with `compute_uv == True`, but both `u` and `v` are not used, then it must be substituted with a new `SVD` `Op` with `compute_uv == False`.

Reply:
Yup that's it!. When you write down the updated rewrite feel free to explain in the comments as much as you did here!

Reply:
There could also be some weird cases where there are 3 SVDs, one with uv and full_matrices that actually doesn't use the uv, and one with uv and not full matrices that actually uses them (or vice-versa). In that case we could replace one for the other, but perhaps that's too much to worry and unlikely to happen. I don't see we ignoring this causing any bug, but we want to at least make sure we don't accidentally rewrite a full-matrices into non full-matrices that are actually used.

Reply:
For this one `return {fgraph.clients[node.outputs[1]]: cl.outputs[0]}` is this the correct syntax?

Reply:
Yup, that tells to replace the key by the value variable